hoodie-client/src/main/java/com/uber/hoodie/io/HoodieAppendHandle.java

```diff
@@ -162,35 +169,59 @@ private void init(Iterator<HoodieRecord<T>> recordItr) {
 
   public void doAppend() {
 
-    List<IndexedRecord> recordList = new ArrayList<>();
-    List<String> keysToDelete = new ArrayList<>();
+    int maxBlockSize = config.getLogFileDataBlockMaxSize();
+    int maxAllowedRecords = (int) (maxBlockSize / averageRecordSize);
     Map<HoodieLogBlock.LogMetadataType, String> metadata = Maps.newHashMap();
     metadata.put(HoodieLogBlock.LogMetadataType.INSTANT_TIME, commitTime);
-    Iterator<HoodieRecord<T>> recordsItr = records.iterator();
-    while (recordsItr.hasNext()) {
-      HoodieRecord record = recordsItr.next();
-      Optional<IndexedRecord> indexedRecord = getIndexedRecord(record);
-      if (indexedRecord.isPresent()) {
-        recordList.add(indexedRecord.get());
-      } else {
-        keysToDelete.add(record.getRecordKey());
+    while (recordItr.hasNext()) {
+      HoodieRecord record = recordItr.next();
+      if(!addAndCheckIfFlushNeeded(record, maxAllowedRecords)) {
+        continue;
       }
-      recordsItr.remove(); //remove entries when IndexedRecord added to new list
+      doAppend(metadata);
+      maxAllowedRecords = (int) (maxBlockSize / averageRecordSize);
     }
+    doAppend(metadata);
+  }
+
+  private void doAppend(Map<HoodieLogBlock.LogMetadataType, String> metadata) {
     try {
       if (recordList.size() > 0) {
         writer = writer.appendBlock(new HoodieAvroDataBlock(recordList, schema, metadata));
```

Comment:
would n't it be better to refresh your estimate from the writer after the first block is logged?

Reply:
That's what is done in https://github.com/uber/hudi/pull/317/files#diff-165908ecbfe7dbae538c00e8157e7be0R219